### PR TITLE
[Blazor] Update npm dependencies

### PR DESCRIPTION
**Pull request by javiercn**
Updates our NPM dependencies and upgrades the auth projects to use webpack5 inline with the rest of the Blazor projects

### CODE REVIEW

File: src/Components/Web.JS/src/Rendering/BrowserRenderer.ts [hunk 1]
```diff
@@ -304,6 +304,10 @@ export class BrowserRenderer {
     }
 
     return true;
+
+    function isBlazorSelectElement(selectElem: HTMLSelectElement | null) : selectElem is BlazorHtmlSelectElement {
+      return !!selectElem  && (deferredValuePropname in selectElem);
```
Comment (SteveSandersonMS) [inline src/Components/Web.JS/src/Rendering/BrowserRenderer.ts:309]:
```suggestion
      return selectElem && (deferredValuePropname in selectElem);
```

File: src/Components/Web.JS/src/Rendering/BrowserRenderer.ts [hunk 2]
```diff
@@ -560,6 +564,8 @@ function stripOnPrefix(attributeName: string) {
   throw new Error(`Attribute should be an event name, but doesn't start with 'on'. Value: '${attributeName}'`);
 }
 
+type BlazorHtmlSelectElement = (HTMLSelectElement & { _blazorDeferredValue?: string })
```
Comment (SteveSandersonMS) [inline src/Components/Web.JS/src/Rendering/BrowserRenderer.ts:567]:
```suggestion
type BlazorHtmlSelectElement = HTMLSelectElement & { _blazorDeferredValue?: string };
```

But if you're defining this type, could we also use it, and remove the references to `selectElem[deferredValuePropname]` since they could now be `selectElem._blazorDeferredValue`?

File: src/Components/WebAssembly/Authentication.Msal/src/Interop/webpack.config.js [hunk 3]
```diff
@@ -1,25 +1,72 @@
...
+
+    ], args.mode !== 'development' ? [] : [
+        // In most cases we want to use external source map files
+        new webpack.SourceMapDevToolPlugin({
+            filename: '[name].js.map',
+            exclude: 'blazor.webview.js',
+        })
+    ]),
+    stats: {
+        //all: true,
```
Comment (SteveSandersonMS) [inline src/Components/WebAssembly/Authentication.Msal/src/Interop/webpack.config.js:66]:
```suggestion
```

File: src/Components/WebAssembly/Authentication.Msal/src/Interop/webpack.config.js [hunk 4]
```diff
@@ -1,25 +1,72 @@
...
+        extensions: ['.ts', '.js'],
+    },
+    devtool: false, // Source maps configured below
+    module: {
+        rules: [{ test: /\.ts?$/, loader: 'ts-loader' }]
+    },
+    entry: {
+        'AuthenticationService': './AuthenticationService.ts',
+    },
+    output: { path: path.join(__dirname, '/dist', args.mode == 'development' ? '/Debug' : '/Release'), filename: '[name].js' },
```
Comment (SteveSandersonMS) [inline src/Components/WebAssembly/Authentication.Msal/src/Interop/webpack.config.js:17]:
```suggestion
    output: { path: path.join(__dirname, 'dist', args.mode == 'development' ? 'Debug' : 'Release'), filename: '[name].js' },
```

`path.join` already inserts the correct platform-specific separators.

File: src/Components/WebAssembly/WebAssembly.Authentication/src/Interop/webpack.config.js [hunk 5]
```diff
@@ -1,25 +1,72 @@
...
+        extensions: ['.ts', '.js'],
+    },
+    devtool: false, // Source maps configured below
+    module: {
+        rules: [{ test: /\.ts?$/, loader: 'ts-loader' }]
+    },
+    entry: {
+        'AuthenticationService': './AuthenticationService.ts',
+    },
+    output: { path: path.join(__dirname, '/dist', args.mode == 'development' ? '/Debug' : '/Release'), filename: '[name].js' },
```
Comment (SteveSandersonMS) [inline src/Components/WebAssembly/WebAssembly.Authentication/src/Interop/webpack.config.js:17]:
```suggestion
    output: { path: path.join(__dirname, 'dist', args.mode == 'development' ? 'Debug' : 'Release'), filename: '[name].js' },
```

File: src/Components/WebAssembly/WebAssembly.Authentication/src/Interop/webpack.config.js [hunk 6]
```diff
@@ -1,25 +1,72 @@
...
+            'process.env.NODE_DEBUG': false,
+            'Platform.isNode': false
+        }),
+        new DuplicatesPlugin({
+            emitErrors: false,
+            emitHandler: undefined,
+            ignoredPackages: undefined,
+            verbose: false
+        }),
+
```
Comment (SteveSandersonMS) [inline src/Components/WebAssembly/WebAssembly.Authentication/src/Interop/webpack.config.js:57]:
```suggestion
```

File: src/Components/WebAssembly/WebAssembly.Authentication/src/Interop/webpack.config.js [hunk 7]
```diff
@@ -1,25 +1,72 @@
...
+    ], args.mode !== 'development' ? [] : [
+        // In most cases we want to use external source map files
+        new webpack.SourceMapDevToolPlugin({
+            filename: '[name].js.map',
+            exclude: 'blazor.webview.js',
+        })
+    ]),
+    stats: {
+        //all: true,
```
Comment (SteveSandersonMS) [inline src/Components/WebAssembly/WebAssembly.Authentication/src/Interop/webpack.config.js:66]:
```suggestion
```

File: src/Components/WebAssembly/WebAssembly.Authentication/src/Microsoft.AspNetCore.Components.WebAssembly.Authentication.csproj [hunk 8]
```diff
@@ -63,29 +64,33 @@
...
-        <AssetKind>All</AssetKind>
-        <AssetMode>All</AssetMode>
-        <AssetRole>Primary</AssetRole>
-        <OriginalItemSpec>@(_InteropBuildOutput->'%(Identity)'))</OriginalItemSpec>
-      </StaticWebAsset>
+    <DefineStaticWebAssets Condition="@(_InteropBuildOutput) != ''"
+      SourceType="Computed"
+      SourceId="$(PackageId)"
+      ContentRoot="$(YarnWorkingDir)dist\$(Configuration)\"
+      BasePath="_content/$(PackageId)"
```
Comment (SteveSandersonMS) [inline src/Components/WebAssembly/WebAssembly.Authentication/src/Microsoft.AspNetCore.Components.WebAssembly.Authentication.csproj:83]: If it's unintentional to have backslashes in some places and forward slashes in others, can they be made consistent?

File: src/Components/WebAssembly/testassets/Wasm.Authentication.Server/Data/Migrations/20210902111722_Initial.cs [hunk 9]
```diff
@@ -147,10 +154,10 @@ protected override void Up(MigrationBuilder migrationBuilder)
...
                 columns: table => new
                 {
-                    LoginProvider = table.Column<string>(maxLength: 128, nullable: false),
-                    ProviderKey = table.Column<string>(maxLength: 128, nullable: false),
-                    ProviderDisplayName = table.Column<string>(nullable: true),
-                    UserId = table.Column<string>(nullable: false)
+                    LoginProvider = table.Column<string>(type: "TEXT", maxLength: 128, nullable: false),
+                    ProviderKey = table.Column<string>(type: "TEXT", maxLength: 128, nullable: false),
+                    ProviderDisplayName = table.Column<string>(type: "TEXT", nullable: true),
+                    UserId = table.Column<string>(type: "TEXT", nullable: false)
```
Comment (SteveSandersonMS) [inline src/Components/WebAssembly/testassets/Wasm.Authentication.Server/Data/Migrations/20210902111722_Initial.cs:160]: These sort of changes seem unrelated to the PR title. For our longer-term tracking, could you rename the PR (or at least the squashed commit message) to something more accurate?

File: src/Components/WebAssembly/testassets/Wasm.Authentication.Server/Data/Migrations/20210902111722_Initial.cs [hunk 10]
```diff
@@ -222,8 +229,7 @@ protected override void Up(MigrationBuilder migrationBuilder)
                         name: "FK_UserPreferences_AspNetUsers_ApplicationUserId",
                         column: x => x.ApplicationUserId,
                         principalTable: "AspNetUsers",
-                        principalColumn: "Id",
-                        onDelete: ReferentialAction.Restrict);
+                        principalColumn: "Id");
```
Comment (SteveSandersonMS) [inline src/Components/WebAssembly/testassets/Wasm.Authentication.Server/Data/Migrations/20210902111722_Initial.cs:232]: I assume these changes are auto-generated (in which case great), otherwise I don't know why we'd be removing details like the `onDelete` here.

File: src/ProjectTemplates/BlazorTemplates.Tests/BlazorServerTemplateTest.cs [hunk 11]
```diff
@@ -150,7 +150,7 @@ private async Task TestBasicNavigation(Project project, IPage page)
 
             await page.WaitForSelectorAsync("nav");
             // <title> element gets project ID injected into it during template execution
-            Assert.Equal(project.ProjectName.Trim(), (await page.GetTitleAsync()).Trim());
+            Assert.Equal("Index", (await page.GetTitleAsync()).Trim());
```
Comment (SteveSandersonMS) [inline src/ProjectTemplates/BlazorTemplates.Tests/BlazorServerTemplateTest.cs:153]: Cool

File: src/ProjectTemplates/BlazorTemplates.Tests/BlazorWasmTemplateTest.cs [hunk 12]
```diff
@@ -319,36 +319,6 @@ private async Task BlazorWasmHostedTemplate_IndividualAuth_Works(BrowserKind bro
             }
         }
 
-        [Theory]
-        [InlineData(BrowserKind.Chromium, Skip = "https://github.com/dotnet/aspnetcore/issues/28596")]
-        public async Task BlazorWasmStandaloneTemplate_IndividualAuth_Works(BrowserKind browserKind)
```
Comment (SteveSandersonMS) [inline src/ProjectTemplates/BlazorTemplates.Tests/BlazorWasmTemplateTest.cs:324]: Could you clarify why we're removing this test case?

File: src/ProjectTemplates/BlazorTemplates.Tests/BlazorWasmTemplateTest.cs [hunk 13]
```diff
@@ -450,9 +420,7 @@ protected async Task BuildAndRunTest(string appName, Project project, BrowserKin
         private async Task TestBasicNavigation(string appName, IPage page, bool usesAuth = false, bool skipFetchData = false)
         {
             await page.WaitForSelectorAsync("nav");
-
-            // <title> element gets project ID injected into it during template execution
-            Assert.Equal(appName.Trim(), (await page.GetTitleAsync()).Trim());
+            Assert.Equal(usesAuth ? appName.Trim() : "Index", (await page.GetTitleAsync()).Trim());
```
Comment (SteveSandersonMS) [inline src/ProjectTemplates/BlazorTemplates.Tests/BlazorWasmTemplateTest.cs:423]: This seems wrong. Why would the page title depend on whether the project uses auth? Is the template with auth not correctly updated to include the `<PageTitle>` components?

Comment (javiercn) [reply src/ProjectTemplates/BlazorTemplates.Tests/BlazorWasmTemplateTest.cs:423]: There might be a race condition here between the title by default and what gets applied by the head component (I think). 

I'm trying to get the test to just pass since it was broken before and its not relevant to this update (and we don't have to be functionally testing the head component on the templates)

I'm happy to file an issue if we want to follow up on this, however I don't want this PR to become let's fix these tests.

File: src/ProjectTemplates/BlazorTemplates.Tests/BlazorWasmTemplateTest.cs (hunk 12 above)
Comment (javiercn) [reply src/ProjectTemplates/BlazorTemplates.Tests/BlazorWasmTemplateTest.cs:324]: Because it is problematic as it doesn't use valid data and as a result the user is redirected to `https://login.microsoftonline.com`.

File: src/Components/WebAssembly/testassets/Wasm.Authentication.Server/Data/Migrations/20210902111722_Initial.cs (hunk 10 above)
Comment (javiercn) [reply src/Components/WebAssembly/testassets/Wasm.Authentication.Server/Data/Migrations/20210902111722_Initial.cs:232]: Yep, just removed the migrations and recreated them

File: src/Components/WebAssembly/testassets/Wasm.Authentication.Server/Data/Migrations/20210902111722_Initial.cs (hunk 9 above)
Comment (javiercn) [reply src/Components/WebAssembly/testassets/Wasm.Authentication.Server/Data/Migrations/20210902111722_Initial.cs:160]: I had to fix this as part of my testing. Migrations EF weren't agreeing with me

File: src/Components/WebAssembly/WebAssembly.Authentication/src/Microsoft.AspNetCore.Components.WebAssembly.Authentication.csproj (hunk 8 above)
Comment (javiercn) [reply src/Components/WebAssembly/WebAssembly.Authentication/src/Microsoft.AspNetCore.Components.WebAssembly.Authentication.csproj:83]: BasePath explictily uses forward slashes

File: src/ProjectTemplates/BlazorTemplates.Tests/BlazorWasmTemplateTest.cs (hunk 13 above)
Comment (SteveSandersonMS) [reply src/ProjectTemplates/BlazorTemplates.Tests/BlazorWasmTemplateTest.cs:423]: Totally understand not wanting to expand the scope of the PR to include fixing unrelated tests.

However it is dangerous to have assertions that we don't understand, especially relating to possible race conditions, given how sensitive people are about their view that browser automation tests are flaky.

I would rather just remove this assertion entirely than have it be here doing something we don't understand. If you think the assertion is necessary in order for the test to say something worthwhile, then yes please also file an issue saying that someone should investigate what's needed to put back an assertion about the page title.

File: src/ProjectTemplates/BlazorTemplates.Tests/BlazorWasmTemplateTest.cs (hunk 12 above)
Comment (SteveSandersonMS) [reply src/ProjectTemplates/BlazorTemplates.Tests/BlazorWasmTemplateTest.cs:324]: Understood. Would you be able to update the referenced issue #28596 to reflect the fact that it's no longer a *flaky* test, but now is a permanent gap in our test coverage until someone creates a valid test here? Otherwise it will get confusing when we try to look at #28596.

File: src/Components/WebAssembly/WebAssembly.Authentication/src/Microsoft.AspNetCore.Components.WebAssembly.Authentication.csproj (hunk 8 above)
Comment (SteveSandersonMS) [reply src/Components/WebAssembly/WebAssembly.Authentication/src/Microsoft.AspNetCore.Components.WebAssembly.Authentication.csproj:83]: Oh yeah, it's part of a URL. Got it. Thanks.

File: src/ProjectTemplates/BlazorTemplates.Tests/BlazorWasmTemplateTest.cs (hunk 13 above)
Comment (javiercn) [reply src/ProjectTemplates/BlazorTemplates.Tests/BlazorWasmTemplateTest.cs:423]: I agree, however we need to have a way to determine we are ready to continue testing (which is what we used this assertion for)

Comment (javiercn) [reply src/ProjectTemplates/BlazorTemplates.Tests/BlazorWasmTemplateTest.cs:423]: Filed an issue to improve the mechanism https://github.com/dotnet/aspnetcore/issues/36224 will tackle it on a follow-up

File: src/ProjectTemplates/BlazorTemplates.Tests/BlazorWasmTemplateTest.cs (hunk 12 above)
Comment (javiercn) [reply src/ProjectTemplates/BlazorTemplates.Tests/BlazorWasmTemplateTest.cs:324]: I've added some comments to ensure this is covered by CTI somehow.

File: src/ProjectTemplates/BlazorTemplates.Tests/BlazorWasmTemplateTest.cs (hunk 13 above)
Comment (SteveSandersonMS) [reply src/ProjectTemplates/BlazorTemplates.Tests/BlazorWasmTemplateTest.cs:423]: Let's not pick something that's subject to a race condition :)

Comment (SteveSandersonMS) [reply src/ProjectTemplates/BlazorTemplates.Tests/BlazorWasmTemplateTest.cs:423]: > I agree, however we need to have a way to determine we are ready to continue testing

Also, are you sure that's true? It's using `Assert.Equal`, not `Browser.Equal`, so there's nothing "waiting" here, unless I'm misreading it. Am I misunderstanding?

Comment (javiercn) [reply src/ProjectTemplates/BlazorTemplates.Tests/BlazorWasmTemplateTest.cs]: I've fixed this by moving the assert below the one that checks the index page is rendered, that should get rid of the race condition, since at that time the title should already be set

Comment (SteveSandersonMS) [reply src/ProjectTemplates/BlazorTemplates.Tests/BlazorWasmTemplateTest.cs]: Awesome, great solution

**Comment by javiercn** [comment]
🆙📅

**Review by SteveSandersonMS — APPROVED** [review]
Updates look great!

File: eng/targets/Helix.props [hunk 14]
```diff
@@ -20,7 +20,7 @@
     <IsMacHelixQueue Condition="$(HelixTargetQueue.Contains('OSX')) or $(HelixTargetQueue.Contains('macOs'))">true</IsMacHelixQueue>
     <HelixTestName>$(MSBuildProjectName)--$(TargetFramework)</HelixTestName>
     <LoggingTestingDisableFileLogging Condition="'$(IsHelixJob)' == 'true'">false</LoggingTestingDisableFileLogging>
-    <NodeVersion>10.15.3</NodeVersion>
+    <NodeVersion>14.17.6</NodeVersion>
```
Comment (TanayParikh) [inline eng/targets/Helix.props:23]: Let's send out a notification about this as it'll likely require updates to local node installs (https://nodejs.org/en/download/).

File: src/Components/WebAssembly/Authentication.Msal/src/Interop/dist/.gitignore [hunk 15]
```diff
@@ -1,2 +1,3 @@
-*.js
-*.js.map
+**/*.js
```
Comment (TanayParikh) [inline src/Components/WebAssembly/Authentication.Msal/src/Interop/dist/.gitignore:1]: Why are we committing the `dist` directory? Looks to be empty (aside from this) : https://github.com/dotnet/aspnetcore/tree/93a3b43a655af3fc2c0087ec0851f33e8c14d269/src/Components/WebAssembly/Authentication.Msal/src/Interop/dist

File: src/Components/WebAssembly/WebAssembly.Authentication/src/Interop/dist/.gitignore [hunk 16]
```diff
@@ -1,2 +1,3 @@
-*.js
-*.js.map
+**/*.js
```
Comment (TanayParikh) [inline src/Components/WebAssembly/WebAssembly.Authentication/src/Interop/dist/.gitignore:1]: Same as above, can we just gitignore `dist/`?

This is empty too: https://github.com/dotnet/aspnetcore/tree/93a3b43a655af3fc2c0087ec0851f33e8c14d269/src/Components/WebAssembly/WebAssembly.Authentication/src/Interop/dist

File: src/Components/WebAssembly/testassets/Wasm.Authentication.Server/Wasm.Authentication.Server.csproj [hunk 17]
```diff
@@ -17,11 +17,11 @@
     <Reference Include="Microsoft.AspNetCore.Components.WebAssembly.Server" />
     <Reference Include="Microsoft.AspNetCore.Identity.EntityFrameworkCore" />
     <Reference Include="Microsoft.AspNetCore.Authentication.Cookies" />
-    <Reference Include="Microsoft.EntityFrameworkCore.Design" />
     <Reference Include="Microsoft.AspNetCore.Identity.UI" />
     <Reference Include="Microsoft.EntityFrameworkCore.Relational" />
     <Reference Include="Microsoft.EntityFrameworkCore.SQLite" />
     <Reference Include="Microsoft.Extensions.Hosting" />
+    <Reference Include="Microsoft.EntityFrameworkCore.Tools"  />
```
Comment (TanayParikh) [inline src/Components/WebAssembly/testassets/Wasm.Authentication.Server/Wasm.Authentication.Server.csproj:24]: Should we also add `microsoft.aspnetcore.components.webassembly.devserver`?

https://github.com/dotnet/runtime/issues/58667#issuecomment-912898439

**Review by TanayParikh — APPROVED** [review]
LGTM, just some minor items

**Comment by TanayParikh** [comment]
Just tried out this branch locally, and it's seems to have fixed a lot of the random issues I was seeing with the `Wasm.Authentication` testassets! 🥇

File: eng/targets/Helix.props (hunk 14 above)
Comment (javiercn) [reply eng/targets/Helix.props:23]: Nobody should be running 10.15, I'll send an email to the team

File: src/Components/WebAssembly/Authentication.Msal/src/Interop/dist/.gitignore (hunk 15 above)
Comment (javiercn) [reply src/Components/WebAssembly/Authentication.Msal/src/Interop/dist/.gitignore:1]: It is because we have a .gitignore for not checking in build outputs.

File: src/Components/WebAssembly/WebAssembly.Authentication/src/Interop/dist/.gitignore (hunk 16 above)
Comment (javiercn) [reply src/Components/WebAssembly/WebAssembly.Authentication/src/Interop/dist/.gitignore:1]: Same answer as above, JS build outputs are produced in that folder

File: src/Components/WebAssembly/testassets/Wasm.Authentication.Server/Wasm.Authentication.Server.csproj (hunk 17 above)
Comment (javiercn) [reply src/Components/WebAssembly/testassets/Wasm.Authentication.Server/Wasm.Authentication.Server.csproj:24]: nope, this is the server of a hosted app
